include/dsn/internal/message_parser.h

```diff
@@ -37,47 +37,84 @@
 #pragma once
 
 # include <dsn/internal/ports.h>
-# include <dsn/internal/rpc_message.h>
 # include <dsn/internal/singleton.h>
+# include <dsn/internal/task_spec.h>
+# include <dsn/cpp/autoref_ptr.h>
+# include <dsn/cpp/utils.h>
 # include <vector>
 
 namespace dsn 
 {
-    class message_parser
+    class message_reader
     {
     public:
-        template <typename T> static message_parser* create(int buffer_block_size, bool is_write_only)
+        explicit message_reader(int buffer_block_size)
+            : _buffer_occupied(0), _buffer_block_size(buffer_block_size) {}
+        ~message_reader() {}
+
+        // called before read to extend read buffer
+        char* read_buffer_ptr(unsigned int read_next);
+
+        // get remaining buffer capacity
+        unsigned int read_buffer_capacity() const { return _buffer.length() - _buffer_occupied; }
+
+        // called after read to mark data occupied
+        void mark_read(unsigned int read_length) { _buffer_occupied += read_length; }
+
+        // discard read data
+        void truncate_read() { _buffer_occupied = 0; }
+
+    public:
+        dsn::blob       _buffer;
+        unsigned int    _buffer_occupied;
+        unsigned int    _buffer_block_size;
+    };
+
+    class message_parser;
+    typedef ref_ptr<message_parser> message_parser_ptr;
+
+    class message_ex;
+
+    class message_parser : public ref_counter
+    {
+    public:
+        template <typename T> static message_parser* create()
         {
-            return new T(buffer_block_size, is_write_only);
+            return new T();
         }
 
-        template <typename T> static message_parser* create2(void* place, int buffer_block_size, bool is_write_only)
+        template <typename T> static message_parser* create2(void* place)
         {
-            return new(place) T(buffer_block_size, is_write_only);
+            return new(place) T();
         }
 
-        typedef message_parser*  (*factory)(int, bool);
-        typedef message_parser*  (*factory2)(void*, int, bool);
+        typedef message_parser*  (*factory)();
+        typedef message_parser*  (*factory2)(void*);
         
     public:
-        message_parser(int buffer_block_size, bool is_write_only);
         virtual ~message_parser() {}
 
-        // before read
-        void* read_buffer_ptr(int read_next);
-        unsigned int read_buffer_capacity() const;
+        // reset the parser
+        virtual void reset() {}
 
         // after read, see if we can compose a message
         // if read_next returns -1, indicated the the message is corrupted
-        virtual message_ex* get_message_on_receive(unsigned int read_length, /*out*/int& read_next) = 0;
+        virtual message_ex* get_message_on_receive(message_reader* reader, /*out*/ int& read_next) = 0;
+
+        // invoked when create response
+        virtual void on_create_response(message_ex* request_msg, message_ex* response_msg) {}
```

Comment:
we already have a join-point for message create. no virtual function defined here to reduce cost.